extensions/amp-font/0.1/amp-font.js

```diff
@@ -0,0 +1,304 @@
+const DEFAULT_SIZE_ = 'medium';
+
+
+class AmpFont extends AMP.BaseElement {
+
+
+  /** @override */
+  prerenderAllowed() {
+    return true;
+  }
+
+
+  /** @override */
+  firstAttachedCallback() {
+    this.customFontFamily_ = this.element.getAttribute('font-family');
+    this.normalizedFontFamilyName_ =
+        normalizeFontFamily_(this.customFontFamily_);
+    this.customFontWeight_ =
+        this.element.getAttribute('font-weight') || DEFAULT_WEIGHT_;
+    this.customFontStyle_ =
+        this.element.getAttribute('font-style') || DEFAULT_STYLE_;
+    this.customFontVariant_ =
+        this.element.getAttribute('font-variant') || DEFAULT_VARIANT_;
+    this.customFontSize_ =
+        this.element.getAttribute('font-size') || DEFAULT_SIZE_;
+    this.measurements_ = {};
+    this.timeoutHandle_ = null;
+    this.hasTimedOut_ = false;
+
+  }
+
+
+  /** @override */
+  buildCallback() {
+    this.startLoad_();
+  }
+
+
+  /**
+   * Start to download the font if it is already not available.
+   * @private
+   */
+  startLoad_() {
+    var addClassName = this.element.getAttribute('on-loading-class') || '';
+    document.documentElement.classList.add(addClassName);
+    this.beginTimeout_();
+    if( "fonts" in document && false) {
+      this.loadFontWithNativeApis_();
+    } else {
+      this.onBeforeloadFont_();
+      this.loadFont_();
+    }
+  }
+
+
+  /**
+   * If the browser supports font loading events use them to load the font.
+   * @private
+   */
+  loadFontWithNativeApis_() {
+    /* style | variant | weight | size/line-height | family */
+    /* font: italic small-caps bolder 16px/3 cursive; */
+    let fontString = (
+      this.customFontStyle_ + " "
+      + this.customFontVariant_ + " "
+      + this.customFontWeight_ + " "
+      + this.customFontSize_ + " "
+      + this.customFontFamily_);
+    if (document.fonts.check(fontString)) {
+      this.onFontLoadSuccess_();
+    } else {
+      document.fonts.load(fontString).then(() => {
+        this.onFontLoadSuccess_();
+      });
+    }
+  }
+
+
+  /**
+   * Step 1 for loading font on browsers that don't support font loading events.
+   * Creates divs hidden from the viewport and measures dimensions for default
+   * fonts.
+   * @private
+   */
+  onBeforeloadFont_() {
+    var ncustomFontFamily_ = normalizeFontFamily_(this.customFontFamily_);
+    var containerElement = document.createElement('div');
+    containerElement.setAttribute(
+      'id', '-amp-font-container-' + ncustomFontFamily_);
+    style.setStyles(containerElement, {
+      fontSize: '48px',
+      fontVariant: 'normal',
+			left: '-999px',
+      lineHeight: 'normal',
+      margin: 0,
+      padding: 0,
+      position: 'absolute',
+			top: '-999px'
+    });
+    DEFAULT_FONTS_.forEach(font => {
+      let defaultFontElement = document.createElement('div');
+      let nfont = normalizeFontFamily_(font);
+      let defaultFontElementId = '-amp-font-'+ ncustomFontFamily_ + '-' + nfont;
```

Comment:
variable concat seems to be much faster

http://jsperf.com/es6-string-literals-vs-string-concatenation